test/model/compose_test.dart

```diff
@@ -0,0 +1,275 @@
+      assert(lines.removeAt(lines.length - 1) == '',
+        'test error: `expected` should end with a newline');
+      final openingFenceLength = firstLineMatch![1]!.length;
+      assert(RegExp(r'^`{' + openingFenceLength.toString() + r'}$').hasMatch(lines.removeAt(lines.length - 1)),
+        'test error: closing fence not found in `expected`');
+      final content = lines.join('\n');
+      check(wrapWithBacktickFence(content: content, infoString: infoString)).equals(expected);
+    }
+
+    test('single line with no code blocks', () {
+      checkFenceWrap('''
+```
+hello world
+```
+''');
+    });
+
+    test('multiple lines with no code blocks', () {
+      checkFenceWrap('''
+```
+hello
+world
+```
+''');
+    });
+
+    test('three-backtick block', () {
+      checkFenceWrap('''
+````
+hello
+```
+code
+```
+world
+````
+''');
+    });
+
+    test('multiple three-backtick blocks; one has info string', () {
+      checkFenceWrap('''
+````
+hello
+```
+code
+```
+world
+```javascript
+// more code
+```
+````
+''');
+    });
+
+    test('whitespace around info string', () {
+      checkFenceWrap('''
+````
+``` javascript 
+// hello world
+```
+````
+''');
+    });
+
+    test('four-backtick block', () {
+      checkFenceWrap('''
+`````
+````
+hello world
+````
+`````
+''');
+    });
+
+    test('five-backtick block', () {
+      checkFenceWrap('''
+``````
+`````
+hello world
+`````
+``````
+''');
+    });
+
+    test('three-, four-, and five-backtick blocks', () {
+      checkFenceWrap('''
+``````
+```
+hello world
+```
+
+````
+hello world
+````
+
+`````
+hello world
+`````
+``````
+''');
+    });
+
+    test('dangling opening fence', () {
+      checkFenceWrap('''
+`````
+````javascript
+// hello world
+`````
+''');
+    });
+
+    test('code blocks marked by indentation or tilde fences don\'t affect result', () {
+      checkFenceWrap('''
+```
+    // hello world
+
+~~~~~~
+code
+~~~~~~
+```
+''');
+    });
+
+    test('backtick fences may be indented up to three spaces', () {
+      checkFenceWrap('''
+````
+ ```
+````
+''');
+      checkFenceWrap('''
+````
+  ```
+````
+''');
+      checkFenceWrap('''
+````
+   ```
+````
+''');
+      // but at 4 spaces of indentation it no longer counts:
+      checkFenceWrap('''
+```
+    ```
+```
+''');
+    });
+
+    test('fence ignored if info string has backtick', () {
+      checkFenceWrap('''
+```
+```java`script
+hello
+```
+''');
+    });
+
+    test('with info string', () {
+      checkFenceWrap(infoString: 'info', '''
+`````info
+```
+hello
+```
+info
+````python
+hello
+````
+`````
+''');
+    });
+  });
+
+  group('narrowLink', () {
+    test('AllMessagesNarrow', () {
+      final store = eg.store();
+      check(narrowLink(store, const AllMessagesNarrow()))
+        .equals(store.account.realmUrl.resolve('#narrow'));
+      check(narrowLink(store, const AllMessagesNarrow(), nearMessageId: 1))
+        .equals(store.account.realmUrl.resolve('#narrow/near/1'));
```

Comment:
(`resolve` is good in these tests, though, because it lets the expected string start with a `#` which is a helpful cue to the reader.)